File: lab02b.py
#!/usr/bin/env python3
"""
Programming Assignment 2: Part Two
"""


def integral(x):
    """
    Approximate an integral by summing the terms of a series.
    The argument must be in the interval (0,1]
    """
    if not (0 < x and x <= 1):
        raise ValueError('argument must be in the interval (0,1]')
    value = x
    numerator = x
    factorial = 1
    small_enough = 1e-6
    limit = 15
    reason = 'Aborted'
    for n in range(1, limit):
        denominator = factorial * (2 * n + 1)
        term = numerator / denominator
        value += term
        numerator *= -x * x
        factorial *= n
        if abs(term) < small_enough:
            reason = 'Accurate'
            break
    return (value, abs(term), f'{reason} after {n+1} iterations')


def test():
    """
    Test cases.
    """
    # x = 0.352234607
    x = 0.587040260
    (value, error, reason) = integral(x)
    print(f'integral({x}) = {value} +/- {error}')
    print(reason)


test()

File: lab02b.cpp
// Programming Assignment 2: Part Two

#include <iostream>
#include <iomanip>
#include <cmath>
#include <limits>       // For std::numeric_limits

int main() {
    double x = 0.0;
    std::cout << "Enter a value for x (0 < x <= 1): ";
    std::cin >> x;

    if (std::cin.fail()) {
        std::cerr << "Error: Invalid input. Please enter a numeric value for x." << std::endl;
        return 1;
    }

    if (x <= 0 || x > 1) {
        std::cerr << "Error: x must be greater than 0 and less than or equal to 1." << std::endl;
        return 1;
    }

    double I_x = 0.0; // The computed value of I(x)
    double term; // The current term in the series
    int n; // Term index
    const double tolerance = 1E-6; // Stopping condition for the term value
    const int max_iterations = 15; // Maximum number of iterations

    double x_power; // x^(2n+1), initially for n=0
    double factorial; // n! for current n

    bool stopped_due_to_small_term = false;

    // Initialize variables before the loop
    term = x; // The first term (n=0)
    x_power = x; // x^(2n+1), initially for n=0
    factorial = 1.0; // n! for current n

    for (n = 0; n < max_iterations; ++n) {
        I_x += term;

        // Compute the next term in the series
        factorial *= (n + 1); // Update factorial for next n
        x_power *= x * x; // Update x_power for next n: x^(2n+1) = x^(2n-1+1) * x^2 (multiplying by x*x advances the exponent by 2)

        int sign = (n % 2 == 0) ? -1 : 1;
        double numerator = sign * x_power;
        double denominator = factorial * (2 * (n + 1) + 1);
        term = numerator / denominator;

        if (std::abs(term) < tolerance) {
            stopped_due_to_small_term = true;
            break;
        }
    }

    if (stopped_due_to_small_term) {
        std::cout << "Stopped due to small term: " << term << std::endl;
    } else {
        std::cout << "Stopped after reaching maximum iterations: " << max_iterations << std::endl;
    }

    std::cout << std::fixed << std::setprecision(8);
    std::cout << "Computed value of I(" << x << ") = " << I_x << std::endl;
    std::cout << "Last term computed: " << term << std::endl;

    return 0;
}
